driver/mloPdistHost.hpp

```diff
@@ -0,0 +1,84 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#pragma once
+
+#include <math.h>
+
+#include <miopen/miopen.h>
+#include <miopen/pdist/utils.hpp>
+#include <miopen/tensor.hpp>
+#include <miopen/tensor_view_utils.hpp>
+#include <../test/ford.hpp>
+
+template <typename Tgpu, typename Tcheck>
+int32_t mloPdistBackwardRunHost(const miopenTensorDescriptor_t inputDesc,
+                                const miopenTensorDescriptor_t outputDesc,
+                                const miopenTensorDescriptor_t doutputDesc,
+                                const miopenTensorDescriptor_t dinputDesc,
+                                const Tgpu* input,
+                                const Tgpu* output,
+                                const Tgpu* doutput,
+                                Tcheck* dinputHost,
+                                const double p)
+{
+    auto input_tv   = miopen::get_inner_expanded_tv<2>(miopen::deref(inputDesc));
+    auto output_tv  = miopen::get_inner_expanded_tv<1>(miopen::deref(outputDesc));
+    auto doutput_tv = miopen::get_inner_expanded_tv<1>(miopen::deref(doutputDesc));
+    auto dinput_tv  = miopen::get_inner_expanded_tv<2>(miopen::deref(dinputDesc));
+
+    auto input_numel = miopen::deref(inputDesc).GetElementSize();
+    size_t N         = miopen::deref(inputDesc).GetLengths()[0];
+    size_t M         = miopen::deref(inputDesc).GetLengths()[1];
+
+    // Fill dinputHost with zeros
+    std::fill(dinputHost, dinputHost + input_numel, static_cast<Tcheck>(0));
+
+    for(size_t i = 0; i < N - 1; ++i)
+    {
+        for(size_t j = i + 1; j < N; ++j)
+        {
+            size_t k       = j + N * i - i * (i + 1) / 2 - i - 1;
+            float grad_k   = static_cast<float>(doutput[doutput_tv.get_tensor_view_idx({k})]);
+            float output_k = static_cast<float>(output[output_tv.get_tensor_view_idx({k})]);
+
+            for(size_t m = 0; m < M; ++m)
+            {
+                float input_first = static_cast<float>(input[input_tv.get_tensor_view_idx({i, m})]);
+                float input_second =
+                    static_cast<float>(input[input_tv.get_tensor_view_idx({j, m})]);
+                float diff = input_first - input_second;
+
+                Tcheck res =
+                    static_cast<Tcheck>(miopen::pdist::backward<float>(diff, grad_k, output_k, p));
+
+                dinputHost[dinput_tv.get_tensor_view_idx({i, m})] += res;
+                dinputHost[dinput_tv.get_tensor_view_idx({j, m})] -= res;
```

Comment:
solved

Reply:
can I do this instead, to avoid multiple times of converting?
```
Tcheck res =
                    static_cast<Tcheck>(miopen::pdist::backward<float>(diff, grad_k, output_k, p));

dinputHost[dinput_tv.get_tensor_view_idx({i, m})] =+ res;
dinputHost[dinput_tv.get_tensor_view_idx({j, m})] =- res;
```

Because I saw you done something similar here: https://github.com/moreh-dev/MIOpen/blob/0c707e6e1c5c24e5cc3e27da520e009104a890e8/driver/unsortedsegmentsum_driver.hpp#L65

Reply:
UPDATE: ask Hieu-nim again, he said we shouldn't calculate using dtype `Tcheck`. Solved this review